gflownet/envs/crystals/spacegroup.py

```diff
@@ -656,11 +673,92 @@ def build_n_atoms_compatibility_dict(n_atoms: List[int], space_groups: List[int]
         assert all([sg > 0 and sg <= 230 for sg in space_groups])
         return {sg: Group(sg).check_compatible(n_atoms)[0] for sg in space_groups}
 
+    def _restrict_space_groups(self, sg_subset: Optional[Iterable] = None):
+        """
+        Updates the dictionaries:
+            - self.space_groups
+            - self.crystal_lattice_systems
+            - self.point_symmetries
+        by eliminating the space groups that are not in the subset sg_subset passed as
+        an argument.
+        """
+        if sg_subset is None:
+            return
+        sg_subset = set(sg_subset)
+
+        # Update self.space_groups
+        self.space_groups = deepcopy(self.space_groups)
+        sg_to_remove = [sg for sg in self.space_groups if sg not in sg_subset]
+        for sg in sg_to_remove:
+            del self.space_groups[sg]
+
+        # Update self.crystal_lattice_systems based on space groups
+        self.crystal_lattice_systems = deepcopy(self.crystal_lattice_systems)
+        cls_to_remove = []
+        for cls in self.crystal_lattice_systems:
+            cls_space_groups = sg_subset.intersection(
+                set(self.crystal_lattice_systems[cls]["space_groups"])
+            )
+            if len(cls_space_groups) == 0:
+                cls_to_remove.append(cls)
+            else:
+                self.crystal_lattice_systems[cls]["space_groups"] = list(
+                    cls_space_groups
+                )
+        for cls in cls_to_remove:
+            del self.crystal_lattice_systems[cls]
+
+        # Update self.point_symmetries based on space groups
+        self.point_symmetries = deepcopy(self.point_symmetries)
+        ps_to_remove = []
+        for ps in self.point_symmetries:
+            ps_space_groups = sg_subset.intersection(
+                set(self.point_symmetries[ps]["space_groups"])
+            )
+            if len(ps_space_groups) == 0:
+                ps_to_remove.append(ps)
+            else:
+                self.point_symmetries[ps]["space_groups"] = list(ps_space_groups)
+        for ps in ps_to_remove:
+            del self.point_symmetries[ps]
+
+        # Update self.crystal_lattice_systems based on point symmetries
+        cls_to_remove = []
+        point_symmetries = set(self.point_symmetries)
+        for cls in self.crystal_lattice_systems:
+            cls_point_symmetries = point_symmetries.intersection(
+                set(self.crystal_lattice_systems[cls]["point_symmetries"])
+            )
+            if len(cls_point_symmetries) == 0:
+                cls_to_remove.append(cls)
+            else:
+                self.crystal_lattice_systems[cls]["point_symmetries"] = list(
+                    cls_point_symmetries
+                )
+        for cls in cls_to_remove:
+            del self.crystal_lattice_systems[cls]
```

Comment:
I get why updating the crystal lattice systems based on the point symmetries (and inversely, just below) but are there really cases where we do end up removing crystal lattice systems or point symmetries at this stage?
For a CLS to be removed here, we need to have a `sg_subset` such that : 
- it excludes all of the space groups associated with PS compatible with this CLS
- but it includes at least one of the space groups associated with this CLS (otherwise it would have been removed earlier in the function)

But if `sg_subset` includes at least one of the space groups associated with this CLS, then, since that SG does have a PS, there must be a PS compatible with this CLS that is still present at this point in the method.
Or am I missing something?

Reply:
I don't have an illuminating answer right now, but what I can say is that I added this piece of code reactively, after seeing the incorrect CLS and PS remained. In any case, I do think all this could be simplified a lot.

Reply:
In that case, if we think it ought to work without that piece of code but it doesn't... then it makes me a bit suspicious about that code

Reply:
Alright, so I just got the latest version of this branch, added breakpoints as such 
```
        # Update self.crystal_lattice_systems based on point symmetries
        cls_to_remove = []
        point_symmetries = set(self.point_symmetries)
        for cls in self.crystal_lattice_systems:
            cls_point_symmetries = point_symmetries.intersection(
                set(self.crystal_lattice_systems[cls]["point_symmetries"])
            )
            if len(cls_point_symmetries) == 0:
                cls_to_remove.append(cls)
            else:
                self.crystal_lattice_systems[cls]["point_symmetries"] = list(
                    cls_point_symmetries
                )
        for cls in cls_to_remove:
            import pdb; pdb.set_trace()
            del self.crystal_lattice_systems[cls]

        # Update self.point_symmetries based on point symmetries
        ps_to_remove = []
        crystal_lattice_systems = set(self.crystal_lattice_systems)
        for ps in self.point_symmetries:
            ps_crystal_lattice_systems = crystal_lattice_systems.intersection(
                set(self.point_symmetries[ps]["crystal_lattice_systems"])
            )
            if len(ps_crystal_lattice_systems) == 0:
                ps_to_remove.append(ps)
            else:
                self.point_symmetries[ps]["crystal_lattice_systems"] = list(
                    ps_crystal_lattice_systems
                )
        for ps in ps_to_remove:
            import pdb; pdb.set_trace()
            del self.point_symmetries[ps]
```
and ran the tests and no breakpoint ever activated.
Which doesn't mean that this code is useless, only (at the very least) that the tests don't include cases that make this code necessary.

Reply:
ok, so would you just remove these part of the code? Or something else? Personally, I don't mind either way and it's the kind of thing I would not put too much work on _right now_ since it is likely to change with new needs.

Reply:
What makes this complicated is that you mentioned that the code didn't work without this addition. If it wasn't for that, I would have said that we can probably remove that code.
Do you remember what you tried that didn't work without adding this code?

If yes, I think the minimal amount of work would be to add a test case for that scenario. Ensure that it works. And see if it works without this code (which might happen if that function has changed since the moment you tested that scenario) to establish if this piece of code is still necessary. If yes, keep it. If no, nuke it. Then we merge the PR.

If no, that's the annoying part. We can't remove the code without possibly breaking something but we don't know _what_ would break. But if we merge as-is, we're potentially merging bugs because I don't how this piece of code would be useful except if there was a bug elsewhere in that function. 
In that case, I don't think that there is a perfect solution. I guess I'd say to keep the code (to be safe), but do 2 things : 
- Add a comment to document our discussion
- Get an additional set of eyes on this function to increase our confidence that there isn't a weird bug being partially hidden by this piece of code that could bite us down the road.

Reply:
> What makes this complicated is that you mentioned that the code didn't work without this addition.

Ok, yes I said that, but it was also during the rush, where I was trying many things like a crazy person. So it's possible I added _this code_ we are talking about together with other parts that _are_ necessary, and I am talking about everything.

> Do you remember what you tried that didn't work without adding this code?

Nope.

Here is another suggestion: run not only the tests but an actual crystal experiment with the breakpoints. If the breakpoints aren't reached I think it is pretty safe to remove.

Reply:
Seems good: https://wandb.ai/alexhg/crystal-gfn/runs/p5ok2fg3

Reply:
Not sure what the difference is between running an experiment with the breakpoint vs running a tests case with the breakpoint.
But given that we've been looking at this for a while, we got tests, and we aren't even sure anymore that this specific part of the code was the part that was added to fix a problem... At this point, I agree with remove and then merge. We don't have any evidence anymore pointing at this code being necessary.

Reply:
> Not sure what the difference is between running an experiment with the breakpoint vs running a tests case with the breakpoint.

I would say the experiment is more exhaustive than the test, but anyway, we agree. I will remove the useless code and merge if CI is happy.

Reply:
For the sake of registering the "findings": I now know why I was confused. I had thought that the _entire block_ of code had to be removed. So I did so, to find that the tests failed. Then realised that one part is necessary, which is consistent with what I had observed. 

Tests are passed now.

Reply:
Doesn't really matter anymore but my point about running an experiment is that, since the allowed space group is an experiment-level hyperparameter, that function will always be called with the same arguments during a given experiment. So running an entire experiment doesn't seem to test more exhaustively then simply adding a single test case in test_spacegroup.py

Reply:
And related to the findings, I'm sorry. I thought I was clear that it was only these two small sections that I was concerned about : 
```
for cls in cls_to_remove:
    del self.crystal_lattice_systems[cls]
```
and
```
for ps in ps_to_remove:
    del self.point_symmetries[ps]
```
I should have been clearer; it would have saved you some time.

Nonetheless, it's great that this PR has finally been merged!

Reply:
No worries, it was my bad! Thanks a lot for your in-depth reviews!